client/src/components/TaskList.js

```diff
@@ -0,0 +1,62 @@
+import React, { Component } from 'react';
+import Task from './Task';
+import axios from 'axios';
+import CreateTask from './CreateTask';
+
+class TaskList extends Component {
+  
+  state = {
+    tasks: []
+  }
+
+  componentDidMount = () => {
+    axios.get('http://localhost:3001/tasks')
+         .then(response => {
+           this.setState({ tasks: response.data });
+         });
+  }
+
+  showNewData = () => {
+    axios.get('http://localhost:3001/tasks')
+    .then(response => {
+      this.setState({ tasks: response.data });
+    })
+  };
+
+  completedTaskHandler = (id) => {
+    axios.patch( 'http://localhost:3001/tasks/'+ id, {is_completed: true})
+    .then( this.showNewData );
+  }
+
+  deleteTaskHandler = (id) => {
+    axios.delete( 'http://localhost:3001/tasks/'+ id)
+    .then( this.showNewData );
+  }
+
+  render(){
+    // タスク未完了のみレンダーする
+    const tasks = this.state.tasks.sort((a, b) => {   
+      return a.id > b.id ? -1 : 1 }).map(task => {
+      return task.is_completed 
+      ? null
+      : <Task 
+          title={task.title}
+          content={task.content}
+          key={task.id}
+          id={task.id}
+          completedClicked={ () => this.completedTaskHandler(task.id)}
+          deleteClicked={ () => this.deleteTaskHandler(task.id)} 
+          // 新しいデータを表示するメソッドを渡す
+          showNewData={ this.showNewData }
```

Comment:
実機でテストしてないからなんともだけど、bindを使ってあげたほうがいいと思う！

```suggestion
          showNewData={this.showNewData.bind(this)}
```

Reply:
bindを理解出来ていなくて調べてみたのですが、
「bindをする事でonclickなどでメソッドを呼び出した際にthisがundefinedになる事を避けるため」と理解したのですが、なぜここでbindを用いた方がいいのか、どういった場面でbindを使うべきなのかが分かりませんでした。

Reply:
確かに俺もちょっと説明するとなると難しいんだけど...w
単純にここの `this. showNewData` の `this` って、TaskListクラスを指してるはずで、showNewDataっていうfunctionはTaskListクラスに属してるはず。

で前提としてJSの挙動として、 以下の２つの挙動が共通ではない。

```js
obj.method();
---
var method = obj.method;
method();
```

つまりthisを指定しないと、引き渡した先のクラスの `showNewData` っていうfunctionを探して、
undefinedになるってことを言ってるみたい。

なのでここからは個人的な意見にはなるけれど、異なるClassComponentにfunctionを受け渡す時はbindでthisを指定してあげるとコーディング上規約を作れば、しょっぱいミスが生まれなくていいよね、とは思う。

あんまり説明になってないかも。俺ももう少ししっかり調べてみる 👍 

[コンポーネントに関数を渡す – React](https://ja.reactjs.org/docs/faq-functions.html)
[Understanding JavaScript Function Invocation and "this"](https://yehudakatz.com/2011/08/11/understanding-javascript-function-invocation-and-this/)

Reply:
なるほど。
undefinedになる挙動はそう言う事なんですね！

引き渡した先で`showNewData`を呼ぶ際に、`this.props`を用いていて、this.propsの中身をコンソールで確認したら渡す側のクラスの`showNewData`というデータが入っていました。
どういった時に引き渡した先のクラスのデータ（つまりここでの例ではshowNewData(undefined)）を参照するという挙動になってしまうのか、というのに対する答えがなかなか出て来ないです。
実際にundefinedになってしまえば、解決できそうな気もするのですが。。